sys/include/can/can.h

```diff
@@ -85,14 +98,26 @@ typedef uint32_t canid_t;
 /**
  * @brief Controller Area Network frame
  */
+#ifdef MODULE_FDCAN
+struct canfd_frame {
+#else
 struct can_frame {
+#endif
     canid_t can_id;  /**< 32 bit CAN_ID + EFF/RTR/ERR flags */
-    uint8_t can_dlc; /**< frame payload length in byte (0 .. CAN_MAX_DLEN) */
+    uint8_t len; /**< frame payload length in byte (0 .. CAN_MAX_DLEN) */
```

Comment:
why not keep it dlc and have a function that calculate len from dlc and backward 

this change requires all users and drivers to change

Reply:
https://www.kernel.org/doc/html/latest/networking/can.html

seems like they went with a union and a separate fd_can type

Reply:
Because as I explain in my commit message, `can_dlc` is deprecated. It is explained also inside the link you provide:

```
Remark: The len element contains the payload length in bytes and should be used instead of can_dlc.
The deprecated can_dlc was misleadingly named as it always contained the plain payload length
in bytes and not the so called 'data length code' (DLC).
```

My first commit already does that change from `can_dlc` to `len` for whole RIOT components.
It then allows native architecture to use `canfd_frame` from `libsocketcan` which does not allow deprecated `can_dlc`.

Reply:
'struct can_frame' is not deprecated ( and it still contains the dlc field ( that one is deprecated) (in union to len)) 
socket can has a new separate `struct canfd_frame` that does not replace struct can_frame.

esspecially on embedded systems its a good idea to keep separate lean structure definitions around. (every canfd_fame you prepare takes 64 + 8 bytes while a can_frame only takes 16 if you need some can_fd messages but many can_messages there will be a huge amount of "wasted" memory"

there are messages that one just might want to keep around and not generate:
switch motor on 16 bytes or 72 bytes
switch motor off 16 bytes or 72 bytes
emegency stop 16 bytes or 72 bytes
...

before this PR on can have a "struct can_frame" that is working on linux and use it on RIOT 

after this pr "struct can_frame" is not available if "MODULE_FDCAN" is active

Reply:
Good catch, I fixed the PR in this way, being able to use `struct can_frame` when `fdcan` module is enabled.

Reply:
As SocketCAN documentaion (till the link you mentioned before) says:
```
The struct canfd_frame and the existing struct can_frame have the can_id, the payload length and the
payload data at the same offset inside their structures. This allows to handle the different
structures very similar. When the content of a struct can_frame is copied into a struct canfd_frame
all structure elements can be used as-is - only the data[] becomes extended.
```

So a user can use a `struct can_frame` at the price of a cast to `can_frame_t` or `struct canfd_frame`.

The documentation also says:
```
When introducing the struct canfd_frame it turned out that the data length code (DLC) of the
struct can_frame was used as a length information as the length and the DLC has a 1:1 mapping
in the range of 0 .. 8. To preserve the easy handling of the length information the canfd_frame.len
element contains a plain length value from 0 .. 64. So both canfd_frame.len and can_frame.len
are equal and contain a length information and no DLC.
```

So the only risk would be to have a len upper than 8 for a `struct can_frame` but the risk is reasonable as it should be handled by the driver. It was not the case for this PR but just added a fixup commit for this :sweat_smile: 

(I will test it in couple of days as I do not have my nucleo-g431rb now).

Reply:
So i was able to test sending `struct can_frame` by modifying `tests/sys/conn_can/main.c` in that way:
```diff
diff --git a/tests/sys/conn_can/main.c b/tests/sys/conn_can/main.c
index b470a63689..3712c697dc 100644
--- a/tests/sys/conn_can/main.c
+++ b/tests/sys/conn_can/main.c
@@ -138,7 +138,8 @@ static int _send(int argc, char **argv, bool rtr)
         print_usage();
         return 1;
     }
-    can_frame_t frame;
+    struct can_frame frame = { 0 };
+
     int ifnum = strtol(argv[2], NULL, 0);
     if (ifnum >= CAN_DLL_NUMOF) {
         puts("Invalid interface number");
@@ -152,7 +153,7 @@ static int _send(int argc, char **argv, bool rtr)
         frame.can_id = strtoul(argv[3], NULL, 16);
         frame.len = argc - 4;
     }
-    if (frame.len > DEFAULT_CAN_MAX_DLEN) {
+    if (frame.len > CAN_MAX_DLEN) {
         puts("Invalid length");
         return 1;
     }
@@ -166,13 +167,10 @@ static int _send(int argc, char **argv, bool rtr)
             frame.data[i] = strtol(argv[4 + i], NULL, 16);
         }
     }
-#ifdef MODULE_FDCAN
-    frame.flags |= CANFD_BRS | CANFD_FDF;
-#endif
 
     conn_can_raw_t conn;
     conn_can_raw_create(&conn, NULL, 0, ifnum, 0);
-    int ret = conn_can_raw_send(&conn, &frame, 0);
+    int ret = conn_can_raw_send(&conn, (can_frame_t*)&frame, 0);
     if (ret < 0) {
         puts("Error when trying to send");
     }
```

On the nucleo:
```
> test_can send 0 10 20 30
```

On the Pi4:
```
Rx 0x2 bytes with ID=0x10 is_fd=False: 0x20 0x30
```

Reply:
> My first commit already does that change from can_dlc to len for whole RIOT components.
> It then allows native architecture to use canfd_frame from libsocketcan which does not allow deprecated can_dlc.

- but it doesn't do that in users application 
- it does not change hunderts of books written about classic can its frames
- code that just needs classic can
and there dlc is the name for the field that contains the number of bytes (up to 8).

Reply:
done